frame/safe-mode/src/lib.rs

```diff
@@ -0,0 +1,362 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2022 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use frame_support::{
+	pallet_prelude::*,
+	traits::{
+		CallMetadata, Contains, Currency, Defensive, GetCallMetadata, PalletInfoAccess,
+		ReservableCurrency,
+	},
+};
+use frame_system::pallet_prelude::*;
+use sp_runtime::traits::Saturating;
+use sp_std::{convert::TryInto, prelude::*};
+
+mod benchmarking;
+mod mock;
+mod tests;
+
+pub use pallet::*;
+
+type BalanceOf<T> =
+	<<T as Config>::Currency as Currency<<T as frame_system::Config>::AccountId>>::Balance;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(PhantomData<T>);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		/// The overarching event type.
+		type Event: From<Event<Self>> + IsType<<Self as frame_system::Config>::Event>;
+
+		/// Contains all calls that can be dispatched when the safe-mode is enabled.
+		///
+		/// The `SafeMode` pallet is always included and does not need to be added here.
+		type SafeModeFilter: Contains<Self::Call>;
+
+		/// Currency type for this pallet.
+		type Currency: ReservableCurrency<Self::AccountId>;
+
+		/// How long the safe-mode will stay active when enabled with [`Pallet::enable`].
+		#[pallet::constant]
+		type EnableDuration: Get<Self::BlockNumber>;
+		/// How much the safe-mode can be extended by each [`Pallet::extend`] call.
+		///
+		/// This does not impose a hard limit as the safe-mode can be extended multiple times.
+		#[pallet::constant]
+		type ExtendDuration: Get<Self::BlockNumber>;
+
+		#[pallet::constant]
+		type EnableStakeAmount: Get<Option<BalanceOf<Self>>>;
+		#[pallet::constant]
+		type ExtendStakeAmount: Get<Option<BalanceOf<Self>>>;
+
+		type EnableOrigin: EnsureOrigin<Self::Origin>;
+		type ExtendOrigin: EnsureOrigin<Self::Origin>;
+		type DisableOrigin: EnsureOrigin<Self::Origin>;
+		type RepayOrigin: EnsureOrigin<Self::Origin>;
+
+		// Weight information for extrinsics in this pallet.
+		//type WeightInfo: WeightInfo;
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// The safe-mode is (already) enabled.
+		IsEnabled,
+
+		/// The safe-mode is (already) disabled.
+		IsDisabled,
```

Comment:
```suggestion
		/// The safe-mode is (already) enabled.
		IsAlreadyEnabled,

		/// The safe-mode is (already) disabled.
		IsAlreadyDisabled,
```

This would make these variants sound more like errors

Reply:
I also thought about this, but it can also happen in the `repay_stake` function, which is more of like an `IsStillEnabled`.  
I could also add both, but that seems overkill.